src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java

```diff
@@ -186,17 +189,14 @@ public String url() {
 
     /**
      * Flag for whether or not the build has completed somehow.
-     * Non-null soon after the build starts or is reloaded from disk.
-     * Recomputed in {@link #onLoad} based on {@link FlowExecution#isComplete}.
-     * TODO may be better to make this a persistent field.
-     * That would allow the execution of a completed build to be loaded on demand (JENKINS-45585), reducing overhead for some operations.
-     * It would also remove the need to null out {@link #execution} merely to force {@link #isInProgress} to be false
-     * in the case of broken or hard-killed builds which lack a single head node.
+     * This was previously a transient field, so we may need to recompute in {@link #onLoad} based on {@link FlowExecution#isComplete}.
      */
-    private transient AtomicBoolean completed;
+    AtomicBoolean completed;  // Non-private for testing
+
+    private transient Object logCopyGuard = new Object();
 
     /** map from node IDs to log positions from which we should copy text */
-    private Map<String,Long> logsToCopy;
+    Map<String,Long> logsToCopy;
```

Comment:
Probably worth a `// Non-private for testing` comment, but not strictly necessary since that's basically the only reason anything we do ever is package-private.